cpp/src/neighbors/detail/cagra/cagra_search.cuh

```diff
@@ -108,6 +111,114 @@ void search_main_core(raft::resources const& res,
   }
 }
 
+/**
+ * @brief Performs ANN search using brute force when filter sparsity exceeds a specified threshold.
+ *
+ * This function switches to a brute force search approach to improve recall rate when the
+ * `sample_filter` function filters out a high proportion of samples, resulting in a sparsity level
+ * (proportion of unfiltered samples) exceeding the specified `params.threshold_to_bf`.
+ *
+ * @tparam T data element type
+ * @tparam IdxT type of database vector indices
+ * @tparam internal_IdxT during search we map IdxT to internal_IdxT, this way we do not need
+ * separate kernels for int/uint.
+ *
+ * @param[in] handle
+ * @param[in] params configure the search
+ * @param[in] strided_dataset CAGRA strided dataset
+ * @param[in] metric distance type
+ * @param[in] queries a device matrix view to a row-major matrix [n_queries, index->dim()]
+ * @param[out] neighbors a device matrix view to the indices of the neighbors in the source dataset
+ * [n_queries, k]
+ * @param[out] distances a device matrix view to the distances to the selected neighbors [n_queries,
+ * k]
+ * @param[in] sample_filter a device filter function that greenlights samples for a given query
+ *
+ * @return true If the brute force search was applied successfully.
+ * @return false If the brute force search was not applied.
+ */
+template <typename T,
+          typename InternalIdxT,
+          typename CagraSampleFilterT,
+          typename IdxT      = uint32_t,
+          typename DistanceT = float>
+bool search_using_brute_force(
+  raft::resources const& res,
+  const search_params& params,
+  const strided_dataset<T, IdxT>& strided_dataset,
+  cuvs::distance::DistanceType metric,
+  raft::device_matrix_view<const T, int64_t, raft::row_major> queries,
+  raft::device_matrix_view<InternalIdxT, int64_t, raft::row_major> neighbors,
+  raft::device_matrix_view<DistanceT, int64_t, raft::row_major> distances,
+  CagraSampleFilterT& sample_filter)
+{
+  if (params.threshold_to_bf >= 1.0) { return false; };
+
+  auto n_queries = queries.extent(0);
+  auto n_dataset = strided_dataset.n_rows();
+
+  auto bitset_filter_view = sample_filter.bitset_view_;
```

Comment:
What happens here if the 2d bitmap isn't able to be converted to a 1d bitet without losing information?